src/main/java/com/aws/greengrass/logmanager/LogManagerService.java

```diff
@@ -524,35 +554,36 @@ private void processCloudWatchAttemptLogInformation(Map<String, Set<LogFile>> co
-                + cloudWatchAttemptLogFileInformation.getStartPosition()) {
+        long bytesUploaded = cloudWatchAttemptLogFileInformation.getBytesRead()
+                + cloudWatchAttemptLogFileInformation.getStartPosition();
+
+        if (!logFileGroup.isActiveFile(file) && file.length() == bytesUploaded) {
             Set<LogFile> completedFiles = completedLogFilePerComponent.getOrDefault(componentName,
                     new HashSet<>());
             completedFiles.add(file);
             completedLogFilePerComponent.put(componentName, completedFiles);
-            if (currentProcessingLogFilePerComponent.containsKey(componentName)) {
-                CurrentProcessingFileInformation fileInformation = currentProcessingLogFilePerComponent
-                        .get(componentName);
-                if (fileInformation.fileHash.equals(fileHash)) {
-                    currentProcessingLogFilePerComponent.remove(componentName);
-                }
-            }
-        } else {
-            // Add the file to the current processing list for the component.
-            // Note: There should always be only 1 file which will be in progress at any given time.
-            // @deprecated The fileName is deprecated value in versions greater than 2.2, but keep it here to avoid
-            // upgrade-downgrade issues.
-            CurrentProcessingFileInformation processingFileInformation =
-                    CurrentProcessingFileInformation.builder()
-                            .fileName(file.getSourcePath())
-                            .startPosition(cloudWatchAttemptLogFileInformation.getStartPosition()
-                                    + cloudWatchAttemptLogFileInformation.getBytesRead())
-                            .lastModifiedTime(cloudWatchAttemptLogFileInformation.getLastModifiedTime())
-                            .fileHash(fileHash)
-                            .build();
-            currentProcessingLogFilePerComponent.put(attemptLogInformation.getComponentName(),
-                    processingFileInformation);
         }
+
+
+        // Track each file that got uploaded. So that if modified in the future we can still track upload
+        // the new contents starting from the last upload position.
+
+        ProcessingFileInformation processingFileInformation =
+                ProcessingFileInformation.builder()
+                        .fileName(file.getSourcePath()) // @deprecated
+                        .startPosition(bytesUploaded)
+                        .lastModifiedTime(cloudWatchAttemptLogFileInformation.getLastModifiedTime())
+                        .fileHash(fileHash)
+                        .build();
+
+        if (!processingFileInformationNELSON.containsKey(componentName)) {
+            // TODO: need to figure a smarter mechanism to adjust the LRU capacity. Add an adjust capacity method
+            processingFileInformationNELSON.put(componentName, new ProcessingFileLRU(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Your code calls `ConcurrentHashMap.containsKey`. If `containsKey` returns `false`, your code then calls `ConcurrentHashMap.put`. Because these operations are not atomic, the key could already exist in the map when you call `put()`, even if `containsKey` returns `false`. This could lead to data loss. Instead of `put()`, call `putIfAbsent()`, which associates an unassociated key with a value atomically. See the following diffs on GitHub that make this code change: [diff1](https://github.com/ninjaframework/ninja/commit/921cbd8b05a4ea930a5a5ad7e451a53e15ec03b7#diff-f39d2fcbd1ca7bbafe398d2a9aaeaf5edd8a23661c40b2f82351ff79ac1b23aeL121), [diff2](https://github.com/kieker-monitoring/kieker/commit/43c9ebd98a4597ea71021c96d04e125ee589d7f8#diff-053788507d3e6fdb69e8ca94e5b2bd0a86b03e3cd2d53a452e18dda69ea071d0L152)

Learn more: [ConcurrentHashMap.putIfAbsent in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#putIfAbsent-K-V-)